refactor prepare_for_backward

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54981 refactor autograd_hook
* **#54977 refactor prepare_for_backward**

put part of codes in prepare_for_backward into functions, so that those functions can be used in static graph training and delay all reduce later on.

Differential Revision: [D27439195](https://our.internmc.facebook.com/intern/diff/D27439195/)

zhaojuanmao committed:
refactor prepare_for_backward

put part of codes in prepare_for_backward into functions, so that those functions can be used in static graph training and delay all reduce later on.

Differential Revision: [D27439195](https://our.internmc.facebook.com/intern/diff/D27439195/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor prepare_for_backward"


put part of codes in prepare_for_backward into functions, so that those functions can be used in static graph training and delay all reduce later on.

Differential Revision: [D27439195](https://our.internmc.facebook.com/intern/diff/D27439195/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1008,6 +996,31 @@ void Reducer::prepare_for_backward(
         "has flow control causing later iterations to have unused parameters.");
   }
 }
+// Traverse the autograd graph starting at the specified output.
```
Should this comment go before search_unused_parameters function now? And leave a more general comment about prepare for backward here?

rohan-varma reviewed (approved):
Lgtm!

zhaojuanmao committed:
Update on "refactor prepare_for_backward"


put part of codes in prepare_for_backward into functions, so that those functions can be used in static graph training and delay all reduce later on.

Differential Revision: [D27439195](https://our.internmc.facebook.com/intern/diff/D27439195/)

[ghstack-poisoned]
